spec/std/wait_group_spec.cr

```diff
@@ -0,0 +1,159 @@
+require "spec"
+require "wait_group"
+
+private def block_until_pending_waiter(wg)
+  while wg.@waiting.empty?
+    Fiber.yield
+  end
+end
+
+describe WaitGroup do
+  describe "#add" do
+    it "can't decrement to a negative counter" do
+      wg = WaitGroup.new
+      wg.add(5)
+      wg.add(-3)
+      expect_raises(RuntimeError, "Negative WaitGroup counter") { wg.add(-5) }
+    end
+
+    it "resumes waiters when reaching negative counter" do
+      wg = WaitGroup.new(1)
+      spawn do
+        block_until_pending_waiter(wg)
+        wg.add(-2)
+      rescue RuntimeError
+      end
+      expect_raises(RuntimeError, "Negative WaitGroup counter") { wg.wait }
+    end
+  end
+
+  describe "#done" do
+    it "can't decrement to a negative counter" do
+      wg = WaitGroup.new
+      wg.add(1)
+      wg.done
+      expect_raises(RuntimeError, "Negative WaitGroup counter") { wg.done }
+    end
+
+    it "resumes waiters when reaching negative counter" do
+      wg = WaitGroup.new(1)
+      spawn do
+        block_until_pending_waiter(wg)
+        wg.done
+        wg.done
```

Comment:
Doesn't this spec have a race condition where the waiter could be resumed after the first `wg.done` without raising, if these specs are run under a multithreaded scheduler?

Reply:
There are probably a lot of specs that have race conditions like this. And it's hard to protect them properly because that would require using the primitives that are being tested.
So I think it might be best to ensure the spec runs in a single-threaded scheduler.

Reply:
Damn, you're right, _and_ the test is wrong: the first `#done` is the one that resumes the waiting fiber. I guess I should manipulate the counter back to `0` then call `#done`.